### PR TITLE
Issue #49: Require media view access for Collabora operations.

### PR DESCRIPTION
**Pull request by donquixote**
(Replaced by #12)

See:
- Issue on main repo: https://github.com/CollaboraOnline/collabora-drupal/issues/49
- PR on main repo: https://github.com/CollaboraOnline/collabora-drupal/pull/50

### CODE REVIEW

File: collabora_online.module [hunk 1]
```diff
@@ -123,26 +124,55 @@ function collabora_online_entity_operation(EntityInterface $entity) {
  * Checks access for the new media operations provided by this module.
  */
 function collabora_online_media_access(MediaInterface $media, string $operation, AccountInterface $account): AccessResultInterface {
+    // Use switch () to align with other entity access hooks.
```
Comment (brummbar) [inline collabora_online.module:127]: Not sure about this, any examples? I only found scenarios where they actually run calculations.

File: collabora_online.module [hunk 2]
```diff
@@ -123,26 +124,55 @@ function collabora_online_entity_operation(EntityInterface $entity) {
...
+    $media_view_access = $media->access('view', $account, TRUE);
+    if (!$media_view_access->isAllowed()) {
+        $reason = "Media 'view' access denied.";
+        if ($media_view_access instanceof AccessResultReasonInterface) {
+            $media_view_reason = $media_view_access->getReason();
+            if ($media_view_reason !== NULL) {
+                $reason = "Media 'view' access denied: $media_view_reason";
+            }
+        }
+        return AccessResult::forbidden($reason)
```
Comment (brummbar) [inline collabora_online.module:147]: Returning forbidden should be prevented in module-level code. This because it prevents users to override this behaviour. The ::andIf() method will always return forbidden when one module explicitly forbid an operation.
This prevents extending access checks from other modules or business-specific code.
We should return neutral access in this scenario, and if nowhere else access is granted, it will be prevented.

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+        // the 'view media' permission to anonymous and authenticated roles.
+        // In a kernel test this does not happen.
+        $this->assertSame([], Role::load(RoleInterface::ANONYMOUS_ID)->getPermissions());
+        $this->assertSame([], Role::load(RoleInterface::AUTHENTICATED_ID)->getPermissions());
+
+        // The first user created in a test method is not user 1.
+        $user = $this->createUser();
+        $this->assertSame(2, (int) $user->id());
+        $this->assertFalse($user->hasPermission('view media'));
+    }
```
Comment (brummbar) [inline tests/src/Kernel/CollaboraMediaAccessTest.php:79]: This costs a full install. If we really have to do this, we can merge it in another test method.

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+                'edit any document in collabora',
+                'edit own document in collabora',
+            ]),
+            'not the author - collabora' => $this->createUser([
+                // Grant all relevant media permissions.
+                'view media',
+                'view own unpublished media',
+                'edit own document in collabora',
+                // Require ownership to edit in collabora.
+                'edit own document in collabora',
```
Comment (brummbar) [inline tests/src/Kernel/CollaboraMediaAccessTest.php:164]: One of this should be any, correct?

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+            ]),
+            'not the author - media' => $this->createUser([
+                // Require ownership for regular media view.
+                'view own unpublished media',
+                // Grant all Collabora permissions, even redundant ones.
+                'preview document in collabora',
+                'edit any document in collabora',
+                'edit own document in collabora',
+            ]),
+            'not the author - collabora' => $this->createUser([
```
Comment (brummbar) [inline tests/src/Kernel/CollaboraMediaAccessTest.php:158]: The naming of these two scenario is not clear enough and requires too much effort to understand what is being tested.
In the first we are testing a non-author, no view access scenario. The second one a scenario where a user can only see its own unpublished documents and has only edit in collabora permissions.

File: tests/src/Kernel/CollaboraMediaAccessTest.php [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+        $this->assertEntityAccess(
+            [],
+            $accounts,
+            $media_entities,
+            [
+                'view',
+                'preview in collabora',
+                'edit in collabora',
+            ],
+            negate: TRUE,
```
Comment (brummbar) [inline tests/src/Kernel/CollaboraMediaAccessTest.php:109]: This is a shortcut to avoid to write all the permission results, and I don't like it at all.
We are already using an atypic yet useful approach to testing permissions/access, but let's try to keep it sane.
Remove the negate and explicitly write all the operations and all the routes.

**Review by brummbar — CHANGES_REQUESTED** [review]

File: collabora_online.module (hunk 1 above)
Comment (donquixote) [reply collabora_online.module:127]: Wow, now `node_node_access()` is the only example I can find!
Others use just plain `if (..)`.

File: collabora_online.module (hunk 2 above)
Comment (donquixote) [reply collabora_online.module:147]: I would agree in general.
Let me explain my motivation in this case, then we can still drop it.

So there are two options:
- We return forbidden here. Then no other module can grant Collabora access for a media that the user cannot view.
- We return neutral. Then any other module that implements this hook to grant Collabora access to media also needs to check for media view access.

So:
With option A we close a door on a theoretical edge case, and we are inconsistent with other entity access hook implementations.
With option B, we slightly inconvenience other module developers, and/or we risk that other module developers forget to check for node access.

Also note that the "forbidden" is based on a fully determined media view access check, where all hooks for media view have already fired - just not for Collabora.

The conclusion could still be that option B is more future-proof and closer to conventional behavior.

File: collabora_online.module (hunk 1 above)
Comment (donquixote) [reply collabora_online.module:127]: Btw I am never sure if `in_array()` is really the best solution for this.

File: tests/src/Kernel/CollaboraMediaAccessTest.php (hunk 3 above)
Comment (donquixote) [reply tests/src/Kernel/CollaboraMediaAccessTest.php:79]: It's a kernel test. It's shouldn't be that bad.
But, now that I am testing it, these kernel tests actually take surprisingly long in my local, up to 6 seconds each.
So we can merge it.

File: tests/src/Kernel/PermissionTest.php [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\Tests\media\Traits\MediaTypeCreationTrait;
+use Drupal\Tests\user\Traits\UserCreationTrait;
+use Drupal\user\PermissionHandlerInterface;
+
+/**
+ * Tests dynamically created permissions.
+ */
+class PermissionTest extends KernelTestBase {
```
Comment (donquixote) [inline tests/src/Kernel/PermissionTest.php:24]: Some thoughts about this one @brummbar 
- We could drop the functional test.
- We could move this change to a separate PR, because it is not really required or even related.

For me, the ideal thing would be to have multiple PRs per issue, so that we can isolate changes, but we don't have to open new Jira issues for every technical change.

Comment (brummbar) [reply tests/src/Kernel/PermissionTest.php:24]: No, no multiple PRs. Our tickets have a scope, we need to stay focused on that. If and when we need multiple PRs, it will be discussed in the ticket or between us case by case.

File: collabora_online.module (hunk 2 above)
Comment (brummbar) [reply collabora_online.module:147]: We should always leave open the choice for this. So always neutral.

**Comment by donquixote** [comment]
Closing in favour of:
- #12 

This is needed to change the branch name, and also to keep this PR as an archive of the older solution.
